Sources/GRPCCore/Server.swift

```diff
@@ -0,0 +1,428 @@
+    case notStarted
+    /// The server is starting but isn't accepting requests yet. Can transition to `running`
+    /// and `stopping`.
+    case starting
+    /// The server is running and accepting RPCs. Can transition to `stopping`.
+    case running
+    /// The server is stopping and no new RPCs will be accepted. Existing RPCs may run to
+    /// completion. May transition to `stopped`.
+    case stopping
+    /// The server has stopped, no RPCs are in flight and no more will be accepted. This state
+    /// is terminal.
+    case stopped
+  }
+
+  private let state: LockedValueBox<State>
+
+  /// Creates a new server with no resources.
+  ///
+  /// You can add resources to the server via ``transports-swift.property``,
+  /// ``services-swift.property``, and ``interceptors-swift.property`` and start the server by
+  /// calling ``run()``. Any changes to resources after ``run()`` has been called will be ignored.
+  public init() {
+    self.transports = Transports()
+    self.services = Services()
+    self.interceptors = Interceptors()
+    self.state = LockedValueBox(.notStarted)
+  }
+
+  /// Starts the server and runs until all registered transports have closed.
+  ///
+  /// No RPCs are processed until all transports are listening. If a transport fails to start
+  /// listening then all open transports are closed and a ``ServerError`` is thrown.
+  ///
+  /// This function returns when all transports have stopped listening and all requests have been
+  /// handled. You can signal to transports that they should stop listening by calling
+  /// ``stopListening()``. The server will continue to process existing requests.
+  ///
+  /// To stop the server more abruptly you can cancel the task that this function is running in.
+  ///
+  /// You must register all resources you wish to use with the server before calling this function
+  /// as changes made after calling ``run()`` won't be reflected.
+  ///
+  /// - Note: You can only call this function once, repeated calls will result in a
+  ///   ``ServerError`` being thrown.
+  public func run() async throws {
+    try self.state.withLockedValue { state in
+      switch state {
+      case .notStarted:
+        state = .starting
+
+      case .starting, .running:
+        throw ServerError(
+          code: .serverIsAlreadyRunning,
+          message: "The server is already running and can only be started once."
+        )
+
+      case .stopping, .stopped:
+        throw ServerError(
+          code: .serverIsStopped,
+          message: "The server has stopped and can only be started once."
+        )
+      }
+    }
+
+    // When we exit this function we must have stopped.
+    defer {
+      self.state.withLockedValue { $0 = .stopped }
+    }
+
+    var listeners: [RPCAsyncSequence<Stream>] = []
+    listeners.reserveCapacity(self.transports.values.count)
+
+    for transport in self.transports.values {
+      do {
+        let listener = try await transport.listen()
```

Comment:
Interesting. I did not expect this. I expected that we had to create a child task per transport and each transport has a `run` method as well? In which task do transports run here?

Reply:
We do have a child task per transport (see `handleRequests`). We start them all in the same task so we can ensure they all start successfully.

Reply:
Just to clarify this. Does that mean that `beginListening` is expected to bind the port? Because so far we always followed the `run` pattern where inside the `run()` method call we bind the ports and then consume the inbound connection/data.

Reply:
`listen()` is expected to bind the port. `listen()` is on the transport which is a lower-level API which users aren't expected to interact with. The `run()` on the server which we do expect users to interact with _does_ do the bind and consume.